### ISSUE TITLE
Libnd4j/SameDiff dropout issues

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#5650]
Related: https://github.com/deeplearning4j/deeplearning4j/issues/5485

Currently, we have DropOut and DropOutInverted ops (the latter is used in DL4J). Also AlphaDropout.

There are some issues with using these in SameDiff:
1. We use p = "input retain probability". TF/Keras uses p = "input drop probability". Do we want to align this (at least at the public API) in SameDiff? (downside: it'd be inconsistent with DL4J)
2. We need to know the dropout mask for peforming backprop through inverted dropout
3. We need some sort of support for 'train vs inference' mode

I'm not sure how we handle (2). But some options:
(a) Decompose dropout into 'generate mask' and 'apply mask' steps. This would work, and would be usable now without anything special; however, it pushes the train vs. inference problem into the general RNG distributions (which isn't ideal - we need to conditionally generate an all 1s array from a binomial distribution at inference time)
(b) Add some sort of stateful storage mechanism for ops (this is basically what cudnn does for dropout) - store the binary mask and reuse for backprop
(c) (variant of 2) Introduce something along the lines of "op outputs that are not SDVariable outputs"
(d) add an extra RNG seed arg to the ops - re-generate the mask on demand, and update the seed only on forward pass (use same seed on backprop)

As for (3) - we might want to build a more general solution (unless we have one already?) as this will come up in other cases too (batch norm, for example). (Obviously, dropout is train-only. Looks like TF just has this as an op arg, and no-ops if it's inference mode. Only real question is the API for this, and updating the op args on the fly before execution).

Edit: more ops that has the same sort of issue (minus train/test) are Choice op and ProbabilisticMerge op. Need to know the mask for performing backprop.

**Comment by raver119** [comment]
b) already implemented, known as Stash at c++ side. however, it's barely ever used in practice.
d) is almost the proper solution.

**Comment by raver119** [comment]
I think this was implemented in libnd4j. Can we close this issue now?

**Comment by AlexDBlack** [comment]
Still an open issue on the SameDiff AFAIK. Assuming it's working at libnd4j level (and should similarly work when executed from java), then next step would be to gradient check dropout ops to verify it's working correctly.
